### ISSUE TITLE
How to use efficient HRL to train ant gather 

### ISSUE THREAD

**Issue by sash-a** [tensorflow/models#10384]
## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/tree/master/research/efficient-hrl

## 2. Describe the feature you request

It seems that ant gather was not included with efficient HRL, even though maze, push and fall are. The code is very oriented to an env that requires a goal, but ant gather has no goal, I am wondering how to configure the ant gather from rllab to work with efficient HRL?

## 3. Additional details

I have added the gather env to the `create_maze_env` method like this:

```
if env_name == 'AntGather':
  e = AntGatherEnv(n_bins=8, sensor_span=2 * math.pi)
  e.reset()
  wrapped_env = gym_wrapper.GymWrapper(e)
  return wrapped_env
```

I have given it a context with meta reward defined like this: `meta/Context.reward_fn = @task/plain_rewards` however it does not train anything like seen in the paper. I am assuming there are some changes that need to be made since it is not a maze style env, what are those changes?

## 4. Are you willing to contribute it? (Yes or No)
 yes

**Comment by ofirnachum** [comment]
Hi, 

The changes you've made so far sound correct and are what I recall doing when running the experiments myself (I don't have the code available to me any more). Other than this, I would suggest the following:

- Ensure that you have updated the gear ranges: "The ant is equivalent to the standard Rllab Ant, except that its gear range is reduced from (−150, 150) to (−30, 30)".
- Ensure that you have the correct termination conditions: "Each episode is terminated either when the ant falls or at 500 steps".
- I would also suggest trying an alternative termination condition, based only on the 500 steps limit (no termination when the ant falls).

**Comment by sash-a** [comment]
Hi

Thanks for the response. Good catch on the gear ranges, I hadn't done that however it's still definitely not training correctly. This is the uvf reward for gather

![image](https://user-images.githubusercontent.com/22861775/143312545-d239c66c-bd14-441f-b317-50cbdc7c50d1.png)

Looks about right, -3.5k is quite high, but the meta reward is all over the show and doesn't seem like it can learn at all.

![image](https://user-images.githubusercontent.com/22861775/143312319-bf877907-9a5c-456c-b701-7fa2562a3bd6.png)

I'm wondering if the plain rewards option I am using in the config is correctly configured? I've simply copied the maze config, changing only the `meta_context_range`, `meta/Context.reward_fn` and the `task/negative_distance/etc` reward function options, is there anything else I should change?

I'm also curious about what `meta_context_range` in the config should be, it looks like in other environments it's just the world bounds so I think `((-11, -11), (11, 11))` for ant gather is correct? (I've changed the activity range to 10)

**Comment by ofirnachum** [comment]
The `meta_context_range` should be irrelevant to the gather task, since you're using plain_rewards. You may even want to set it to ((0, 0), (0, 0)) to avoid confusing the agent.

Another thing that may help is changing the AntGatherEnv to always return done=False (disable all termination). There may be some issue in how the HIRO code handles environment terminations, so this might fix it.

**Comment by sash-a** [comment]
Thanks, I'll give that a try and see if it helps

**Comment by sash-a** [comment]
I tried both of the fixes above and the meta reward still looks the same, hovering between -0.2 and 0.2. I did find another issue, I was using the rllab ant class which includes the cfrc and body com observations, so I switch to using the HIRO included ant. Is there anything you can think of like observation order which might be an issue (it's currently [obs, sensor readings])? 

This time the primitive reward does look quite strange, seems to struggle to learn at all (this run is hiro_orig, previous was hiro_xy):

![image](https://user-images.githubusercontent.com/22861775/143403057-8b212f77-e0c1-43f1-8f53-98f79fb85b97.png)

**Comment by ofirnachum** [comment]
There's nothing else I can think of that you may be missing. It's strange that it doesn't make any progress on the task... best advice I have right now is to just go over your changes again carefully and make sure there isn't anything else obviously wrong. You could also try making the AntGather task easier somehow -- maybe a smaller arena, or provide the apple/bomb coordinates directly instead of passing in the visual sensor reading.

**Comment by sash-a** [comment]
Will try making it easier, thanks for the help :)

While I have you a quick question on the maze observations: in the paper appendix it says "The agent observes qpos,qvel, the current time step t, and the target location." This may be a stupid question, but I can't seem to find the answer, when you say target location do you mean the raw target location, is it normalized, or is it multiplied by something like 0.01? Closest I can find is [here](https://github.com/tensorflow/models/blob/d86d4aa60a0a826bf76db7e3be77d474946c5b53/research/efficient-hrl/run_env.py#L41), which  implies raw location, but `run_env.py` doesn't seem to be used anywhere

**Comment by ofirnachum** [comment]
The target location is the "meta_context" which is defined in the training configs. I believe it provides absolute coordinates.
